mllib/src/main/scala/org/apache/spark/mllib/pmml/export/PMMLTreeModelUtils.scala

```diff
@@ -0,0 +1,282 @@
+
+private[mllib] object PMMLTreeModelUtils {
+
+  val FieldNamePrefix = "field_"
+
+  def toPMMLTree(dtModel: DecisionTreeModel, modelName: String): (TreeModel, List[DataField]) = {
+
+    val miningFunctionType = getPMMLMiningFunctionType(dtModel.algo)
+
+    val treeModel = new org.dmg.pmml.TreeModel()
+      .withModelName(modelName)
+      .withFunctionName(miningFunctionType)
+      .withSplitCharacteristic(TreeModel.SplitCharacteristic.BINARY_SPLIT)
+
+    var (rootNode, miningFields, dataFields, classes) = buildStub(dtModel.topNode, dtModel.algo)
+
+    // adding predicted classes for classification and target field for regression
+    // for completeness
+    dtModel.algo match {
+
+      case Algo.Classification => {
+        miningFields = miningFields :+ new MiningField()
+          .withName(FieldName.create("class"))
+          .withUsageType(FieldUsageType.PREDICTED)
+
+        val dataField = new DataField()
+          .withName(FieldName.create("class"))
+          .withOpType(OpType.CATEGORICAL)
+          .withValues(classes.asJava)
+          .withDataType(DataType.DOUBLE)
+
+        dataFields = dataFields :+ dataField
+      }
+      case Algo.Regression => {
+        val targetField = FieldName.create("target")
+        val dataField = new DataField(targetField, OpType.CONTINUOUS, DataType.DOUBLE)
+        dataFields = dataFields :+ dataField
+
+        miningFields = miningFields :+ new MiningField()
+          .withName(targetField)
+          .withUsageType(FieldUsageType.TARGET)
+      }
+    }
+
+    val miningSchema = new MiningSchema()
+      .withMiningFields(miningFields.asJava)
+
+    treeModel.withNode(rootNode)
+      .withMiningSchema(miningSchema)
+
+    (treeModel, dataFields)
+
+  }
+
+  /** Build a pmml tree stub given the root mllib node. */
+  private def buildStub(
+              rootDTNode: Node,
+              algo: Algo): (PMMLNode, List[MiningField], List[DataField], List[PMMLValue]) = {
+
+    val miningFields = MutableList[MiningField]()
+    val dataFields = mutable.HashMap[String, DataField]()
+    val classes = mutable.MutableList[Double]()
+
+    def buildStubInternal(rootNode: Node, predicate: Predicate): PMMLNode = {
+
+      // get rootPMML node for the MLLib node
+      val rootPMMLNode = createNode(rootNode)
+      rootPMMLNode.withPredicate(predicate)
+
+      var leftPredicate: Predicate = new True()
+      var rightPredicate: Predicate = new True()
+
+      if (rootNode.split.isDefined) {
+        val fieldName = FieldName.create(FieldNamePrefix + rootNode.split.get.feature)
+        val dataField = getDataField(rootNode, fieldName).get
+
+        if (!dataFields.get(dataField.getName.getValue).isDefined) {
+          dataFields.put(dataField.getName.getValue, dataField)
+          miningFields += new MiningField()
+            .withName(dataField.getName)
+            .withUsageType(FieldUsageType.ACTIVE)
+
+        } else if (dataField.getOpType != OpType.CONTINUOUS.value()) {
```

Comment:
remove `.value()` in the end, because the type of both sides are not the same. `getOpType` gets the OpType, while `OpType.CONTINUOUS.value()` gets a string.